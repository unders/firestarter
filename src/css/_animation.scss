@import "palette";

@keyframes highlight {
  0% { background-color: lighten($accent-color, 10%); }
  100% { background-color: $body-color; }
}

@keyframes errorHighlight {
  0% { background-color: lighten($error-color, 10%); }
  100% { background-color: lighten($error-color, 40%) }
}
