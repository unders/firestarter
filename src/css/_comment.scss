@import "ripple";
@import "space";
@import "palette";
@import "placeholder";

@mixin funcbox-comment($font-size-body) {
  .funcbox-comment-heading {
    margin-top: 0;
    font-size: 16px;
    color: $secondary-text-color;
  }
  .funcbox-placeholder {
    font-size: 20px;
    height: 60px;
    padding-right: 20px;
    padding-left: 20px;
  }
  .funcbox-placeholder, .funcbox-comment-form {
    background: $primary-color-text;
    color: $secondary-text-color;
    border: 1px solid rgba(0,0,0,.09);
    border-radius: 3px;
    display: flex;
    align-items: center;
    cursor: text;
  }
  .funcbox-comment-form {
    justify-content: space-between;
    align-items: stretch;
    flex-direction: column;
  }

  .funcbox-placeholder.hide {
    display: none;
  }

  .funcbox-comment-form.hide {
    display: none;
  }

  .funcbox-comment-textarea {
    padding: 20px 20px 0 20px;
    height: 210px;
    font-size: 18px;
    outline: none;
    resize: none;
    //overflow: hidden;
    border: none;

    @include placeholder(lighten($secondary-text-color, 20%))
  }

  .error .funcbox-comment-textarea {
    @include placeholder(lighten($error-color, 20%));
  }

  .funcbox-comment-footer {
    background: $body-color;
    padding-top: 10px;
    padding-bottom: 10px;
    padding-right: 20px;
    display: flex;
    align-items: center;
    justify-content: flex-end;
  }

  .funcbox-comment-submit {
    color: lighten($secondary-text-color, 20%);
    background: $body-color;
    border: none;
    text-transform: uppercase;
    margin-left: 10px;
    font-size: 16px;
    padding: 10px;
    cursor: pointer;
    outline: none;
  }

  .funcbox-comment-submit:focus {
    background: darken($primary-color-text, 5);
  }

  .funcbox-comment-submit:active {
    background: darken($primary-color-text, 5);
  }

  .funcbox-comment-submit.publish {
    color: $accent-color;
    @include ripple();
  }

  .funcbox-comment-submit:hover {
    background: darken($primary-color-text, 7);
  }

  //
  //  COMMENT-LIST
  //
  .funcbox-comment-list {
    padding:0;
    margin:0;
    list-style-type: none;
  }
  .funcbox-comment-item {
      background: $primary-color-text;
      margin-top: 40px;
      padding: 20px 20px;
      border: 1px solid rgba(0,0,0,.09);
      border-radius: 3px;
  }

  .funcbox-comment-item.highlight {
    animation: highlight 1.5s ease-out;
  }

  @keyframes highlight {
    0% { background-color: lighten($accent-color, 10%); }
    100% { background-color: $body-color; }
  }
}



