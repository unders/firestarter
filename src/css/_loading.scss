// https://github.com/tawian/text-spinners/blob/master/spinners.scss
.loading {
  display: inline-block;
  overflow: hidden;
  height: 1.3em;
  margin-top: -0.3em;
  line-height: 1.5em;
  vertical-align: text-bottom;
}

.loading::after {
  display: inline-table;
  white-space: pre;
  text-align: left;
}

.loading::after {
  content: "\A.\A..\A...";
  animation: spin4 2s steps(4) infinite;
}

/* --- Animations --- */
@keyframes spin4  { to { transform: translateY( -6.0em); } }

