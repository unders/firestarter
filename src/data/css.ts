export class CSS {
    static readonly error = "error";
    static readonly hide = "hide";
    static readonly show = "";
    static readonly empty = "";
}
