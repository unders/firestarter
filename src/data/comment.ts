import { CSS as css } from './css';

export class Comment {
   readonly body: string;

    constructor(body: string) {
       this.body = body;
    }
}

export class CommentError {
    body: string;
    err: string;

    constructor(body: string, err: string) {
        this.body = body;
        this.err = err;
    }
}

export class CommentFormWidget {
    placeholder: CommentFormPlaceholder;
    form: CommentForm;
    submit: CommentFormSubmit;
    data: CommentFormData;

    constructor(submitText: string, placeholderText: string) {
        this.placeholder = new CommentFormPlaceholder(placeholderText);
        this.form = new CommentForm();
        this.submit = new CommentFormSubmit(submitText);
        this.data = new CommentFormData();
    }
}

class CommentFormPlaceholder {
    readonly text: string = "Write a comment...";
    klass: string = css.show;
    constructor(text: string) {
        this.text = text;
    }
}

class CommentForm {
    klass: string = css.hide;
}

class CommentFormSubmit {
    readonly title: string = "Publish";
    disable: boolean = false;

    constructor(title: string) {
        this.title = title
    }
}

class CommentFormData {
    [key: string]: string;
    body: string = "";
}

