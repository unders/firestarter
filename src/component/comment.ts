import { Comment } from "../data/comment"
import { Poster} from "../client/client"
import { Dom as dom } from "../dom/dom"

export interface Props {
    root: string;
    listTitle: string;
    client: Poster;
    comments: Comment[];
}

export class Component {
    root: Element;
    form: Form;
    list: List;
    bind: (template: TemplateStringsArray, ...args : any[]) => void;

    constructor(props: Props) {
        const root = document.querySelector(props.root);

        if (root) {
            this.root = root;
            this.bind = dom.bind(this.root);
            this.form = new Form(props.client, new View(root));
            this.list = new List(props.listTitle, props.comments);

            this.render();
        }
    }

    render() {
        if (this.root) {
            this.bind`${[
                this.form.render()
                // this.list.render()
            ]}`;
        }
    }
}

class View {
    root: Element;
    data: Data = new Data();
    body: Body = new Body();
    submit: Submit = new Submit();

    constructor(root: Element) {
        this.root = root;
        this.resetBodySize = this.resetBodySize.bind(this);
    }

    focusBody() {
        const el = this.root.querySelector(".funcbox-textarea") as HTMLTextAreaElement;
        el.focus();
    }

    resetBodySize() {
        const el = this.root.querySelector(".funcbox-textarea") as HTMLTextAreaElement;
        el.style.height = "inherit";
        el.style.height = el.scrollHeight + "px";
    }

    isValid(): boolean {
        return this.body.isValid(this.data.body);
    }

    toJSON(): string {
        return JSON.stringify(this.data);
    }

    disableSubmit(): void {
        this.submit.disable = true;
    }

    enableSubmit(): void {
        this.submit.disable = false;
    }

    reset(): void {
        this.data.body = "";
        this.body.setOk();
        this.submit.disable = false;
        setTimeout(this.resetBodySize, 1);
    }
}
class Data {
    [key: string]: string;
    body: string = "";
}
class Body {
    static readonly ok: string = "";
    static readonly error: string = "error";
    errorKlass: string = "";
    errorMsg: string = "Please, write something.";
    placeholder: string = "Write a comment";

    isValid(value: string): boolean {
       if (value === "") {
           return this.setError();
       } else {
           return this.setOk();
       }
    }

    setError() {
        this.errorKlass = Body.error;
        return false;
    }

    setOk() {
        this.errorKlass = Body.ok;
        return true
    }
}
class Submit {
    readonly title: string = "Send";
    disable: boolean = false;
}


class Form {
    readonly client: Poster;
    readonly html: (template: TemplateStringsArray, ...args : any[]) => void;
    view: View;
    bodyHasGrow: boolean = false;

    constructor(client: Poster, view: View) {
        this.view = view;
        this.client = client;
        this.html = dom.wire(this);
        this.submit = this.submit.bind(this);
        this.saveInput = this.saveInput.bind(this);
        this.removeError = this.removeError.bind(this);
        this.grow = this.grow.bind(this);
        // this.shrink = this.shrink.bind(this);
    }

    saveInput(e: Event) {
        const input = e.target as HTMLInputElement;
        this.view.data[input.name] = input.value.trim();
    }

    removeError(e: Event) {
        if (this.view.body.errorKlass !== Body.ok) {
            this.view.body.errorKlass = Body.ok;
            this.render();
        }
    }

    grow(e: Event) {
        const el = e.target as HTMLTextAreaElement;

        if (!this.bodyHasGrow && el.scrollHeight > el.clientHeight) {
            this.bodyHasGrow = true;
            el.style.height = el.scrollHeight + "px";
        }

        if (this.bodyHasGrow) {
            el.style.height = "inherit";
            el.style.height = el.scrollHeight + "px";
        }
    }

    async submit(e: Event) {
        e.preventDefault();

        this.view.disableSubmit();
        this.render();

        if(!this.view.isValid()) {
            this.view.enableSubmit();
            this.render();
            this.view.focusBody();
            return;
        }
        const req = this.view.toJSON();
        this.view.reset();
        this.render();

        // animate a fast add at the top of list...

        const {json, err} = await this.client.post(req);
        if (err) {
            // animate and replace list item with error message
            if (err.code == 400) {

            }
            console.log(err.code, err.status, err.message, err.value);
        } else {
            // resolve optimistic update
            // update store.
            console.log("json: ", json);
        }
    }

    render() {
        const data = this.view.data;
        const body = this.view.body;
        const submit = this.view.submit;

        return this.html`
            <form   class="funcbox-comment-form"
                    onsubmit=${this.submit}
                    oninput=${this.saveInput}>
                <div class="${['funcbox-comment-group', body.errorKlass].join(' ')}">
                    <textarea   class="funcbox-textarea"
                                oninput="${this.grow}"
                                onkeydown="${this.removeError}"
                                tabindex="1"
                                name='body'
                                placeholder="${body.placeholder}"
                                value="${data.body}"></textarea>
                    <span class="funcbox-textarea-error">${body.errorMsg}</span>
                </div>
                <button class="funcbox-button"
                        tabindex="2"
                        disabled=${submit.disable}>${submit.title}</button>
            </form>`;
    }
}

class List {
    title: string;
    comments: Comment[];
    html: (template: TemplateStringsArray, ...args : any[]) => void;

    constructor(title: string, comments: Comment[]) {
        this.title = title;
        this.html = dom.wire(this);
    }

    click(e: Event) {
        console.log(e);
        console.log(e.target);
    }

    render() {
        return this.html`
            <h3>${this.title}</h3>
            <ul class="comments">${this.comments.map( (comment) => dom.wire(comment)`
                <li onclick="${this.click}">${comment.body}</li>`)}
            </ul>`;
    }
}
