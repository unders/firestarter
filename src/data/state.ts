import { CommentFormWidget } from './comment'

export interface IState {
    setState(callback: (s: State) => {}): void
    getState(): State
}

export class State {
    commentFormWidget: CommentFormWidget;

    static init(): State {
        const s = new State();
        s.commentFormWidget = new CommentFormWidget("Publish", "Write a comment...");
        return s
    }
}




