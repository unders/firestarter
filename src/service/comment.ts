import { Router as router} from "../router"
import { XHR } from "../client/client"
import {IState, State} from "../data/state"
import {CSS as css} from "../data/css"
import {Comment, CommentListItem} from "../data/comment"

export class CommentService {
    private readonly xhr: XHR;
    private readonly comments: Comment[];
    private state: IState;

    constructor(xhr: XHR, comments: Comment[]) {
        this.xhr = xhr;
        this.comments = comments;
    }

    init(state: IState) {
        this.state = state;

        const listItems: CommentListItem[] = [];
        this.comments.forEach((comment) => {
            const listItem = new CommentListItem(comment, "", "", "");
            listItems.push(listItem);
        });

        this.state.getState().commentListWidget.comments = listItems;
    }

    async submitComment(comment: Comment, timeout: number) {
        const wait = this.minWait(timeout);
        const {json, err} = await this.xhr.post(router.commentPath, comment.toJSON());

        if (err) {
            const listItem = new CommentListItem(comment, css.errHighlight, err.header, err.message);

            await wait;
            this.state.setState((state: State): any => {
                state.commentListWidget.comments.unshift(listItem);
            });
            return;
        }

        const listItem = new CommentListItem(comment, css.highlight, "", "");

        await wait;
        this.state.setState((state: State): any => {
            state.commentListWidget.comments.unshift(listItem);
        });
    }

    private minWait(timeout: number): Promise<any> {
        return new Promise<any>((resolve) => {
            if (timeout === 0) {
                resolve();
            } else {
                setTimeout(resolve, timeout);
            }
        });
    }
}
