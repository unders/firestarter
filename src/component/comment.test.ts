import { CSS as css } from '../data/css';
import { CommentService } from '../service/comment';
import { CommentComponent } from './comment';
import { CommentFormWidget, CommentListWidget } from '../data/comment';
import { newClient} from "../test/mock";
import {adjustSnap, sleep} from "../test/helper";
import { State } from "../data/state";
import { XMLHttpRequestMock } from "../test/mock";


class Pager {
    private readonly mock: XMLHttpRequestMock;
    private readonly root: HTMLElement;
    private readonly formWidget: CommentFormWidget;
    private readonly listWidget: CommentListWidget;

    constructor(root: HTMLElement, form: CommentFormWidget, list: CommentListWidget, mock: XMLHttpRequestMock) {
        this.root = root;
        this.formWidget = form;
        this.listWidget = list;
        this.mock = mock
    }

    showForm() {
        const el = this.root.querySelector("[data-showForm]") as HTMLElement;
        expect(el).not.toBeNull();
        el.click();
        expect(this.formWidget.placeholder.klass).toEqual(css.hide);
        expect(this.formWidget.form.klass).toEqual(css.show);
    }
    cancelForm() {
        const el = this.root.querySelector("[data-cancel]") as HTMLElement;
        expect(el).not.toBeNull();
        el.click();
        expect(this.formWidget.placeholder.klass).toEqual(css.show);
        expect(this.formWidget.form.klass).toEqual(css.hide);
        expect(this.formWidget.submit.disable).toBeFalsy();
        expect(this.formWidget.data.body).toEqual("");
    }
    submitInvalidForm() {
        const el = this.root.querySelector("[data-submit]") as HTMLElement;
        expect(el).not.toBeNull();
        el.click();
        expect(this.formWidget.placeholder.klass).toEqual(css.hide);
        expect(this.formWidget.form.klass).toEqual(css.error);
    }
    writeComment(text: string) {
        const el = this.root.querySelector("[data-body]") as HTMLTextAreaElement;
        expect(el).not.toBeNull();
        el.value = text;
        const evt = new Event("input", {
            'bubbles': true,
            'cancelable': true
        });
        el.dispatchEvent(evt);
        expect(this.formWidget.form.klass).toEqual(css.empty);
        expect(this.formWidget.data.body).toEqual(text);
    }
    async submitForm(status: number) {
        const el = this.root.querySelector("[data-submit]") as HTMLElement;
        expect(el).not.toBeNull();
        el.click();
        expect(this.formWidget.form.klass).toEqual(css.hide);
        expect(this.formWidget.placeholder.klass).toEqual(css.show);
        expect(this.formWidget.submit.disable).toBeFalsy();
        expect(this.formWidget.data.body).toEqual("");
        await this.mock.simulateResponse(status, "{}");
    }
    hasHighlightedComment(index: number, want: string) {
        const comment = this.listWidget.comments[index];
        expect(comment.data.body).toEqual(want);
        expect(comment.klass).toEqual(css.highlight);
        expect(comment.errorKlass).toEqual(css.hide);
    }
    hasErrorComment(index: number, want: string) {
        const comment = this.listWidget.comments[index];
        expect(comment.data.body).toEqual(want);
        expect(comment.klass).toEqual(css.errHighlight);
        expect(comment.errorKlass).toEqual(css.show);
    }
}

describe("CommentComponent", () => {
    const [mock, client] = newClient();
    const state = State.newIState();
    const service = new CommentService(client, []);
    service.init(state);
    document.body.innerHTML = '<div id="funcbox-comment"></div>';
    const root = document.querySelector("#funcbox-comment") as HTMLElement;
    CommentComponent.timeout = () => { return 0; };
    const comment = new CommentComponent({ root: root, service: service, state: state });
    const s = state.getState();
    const page = new Pager(root, s.commentFormWidget, s.commentListWidget, mock);

    test("#render()", () => {
        comment.render();
        expect(adjustSnap(document.body.innerHTML)).toMatchSnapshot();
    });

    test("show form", () => {
        page.showForm();
        comment.render();
        expect(adjustSnap(document.body.innerHTML)).toMatchSnapshot();
    });

    test("cancel form", () => {
        page.cancelForm();
        comment.render();
        expect(adjustSnap(document.body.innerHTML)).toMatchSnapshot();
    });

    test("publish form", async () => {
        // Add comment without any text
        page.showForm();
        page.submitInvalidForm();
        comment.render();
        expect(adjustSnap(document.body.innerHTML)).toMatchSnapshot();

        // add valid comment
        page.writeComment("This is a comment text.");
        comment.render();
        await page.submitForm(200);
        comment.render();
        expect(adjustSnap(document.body.innerHTML)).toMatchSnapshot();
        await sleep(1);
        page.hasHighlightedComment(0,"This is a comment text.")

        // add comment that fails on server...
        page.writeComment("This is another comment.");
        comment.render();
        await page.submitForm(500); // crash on server
        comment.render();
        expect(adjustSnap(document.body.innerHTML)).toMatchSnapshot();
        await sleep(1);
        page.hasErrorComment(0, "This is another comment.")
    });

});
