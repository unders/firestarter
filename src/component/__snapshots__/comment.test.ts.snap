// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentComponent #form #PublishForm() shows error when no input data 1`] = `
"<div id=\\"funcbox-comment\\"><h3 class=\\"funcbox-comment-heading\\">Comments</h3><div class=\\"funcbox-placeholder hide\\" data-showform=\\"1\\">
                        <span class=\\"funcbox-placeholder-text\\">
                            Write a comment...
                        </span>
                    </div><form class=\\"funcbox-comment-form error\\">
                        <textarea class=\\"funcbox-comment-textarea\\" name=\\"body\\" placeholder=\\"Write a comment...\\"></textarea>
                        <div class=\\"funcbox-comment-footer\\">
                            <button class=\\"funcbox-comment-submit\\" data-cancel=\\"1\\">Cancel</button>
                            <button class=\\"funcbox-comment-submit publish\\" data-submit=\\"1\\">Publish</button>
                        </div>
                    </form><ul class=\\"funcbox-comment-list\\">
                <!--;-->
            </ul><!--;--></div>"
`;

exports[`CommentComponent #form #cancelForm() 1`] = `
"<div id=\\"funcbox-comment\\"><h3 class=\\"funcbox-comment-heading\\">Comments</h3><div class=\\"funcbox-placeholder \\" data-showform=\\"1\\">
                        <span class=\\"funcbox-placeholder-text\\">
                            Write a comment...
                        </span>
                    </div><form class=\\"funcbox-comment-form hide\\">
                        <textarea class=\\"funcbox-comment-textarea\\" name=\\"body\\" placeholder=\\"Write a comment...\\"></textarea>
                        <div class=\\"funcbox-comment-footer\\">
                            <button class=\\"funcbox-comment-submit\\" data-cancel=\\"1\\">Cancel</button>
                            <button class=\\"funcbox-comment-submit publish\\" data-submit=\\"1\\">Publish</button>
                        </div>
                    </form><ul class=\\"funcbox-comment-list\\">
                <!--;-->
            </ul><!--;--></div>"
`;

exports[`CommentComponent #form #showForm() 1`] = `
"<div id=\\"funcbox-comment\\"><h3 class=\\"funcbox-comment-heading\\">Comments</h3><div class=\\"funcbox-placeholder hide\\" data-showform=\\"1\\">
                        <span class=\\"funcbox-placeholder-text\\">
                            Write a comment...
                        </span>
                    </div><form class=\\"funcbox-comment-form \\">
                        <textarea class=\\"funcbox-comment-textarea\\" name=\\"body\\" placeholder=\\"Write a comment...\\"></textarea>
                        <div class=\\"funcbox-comment-footer\\">
                            <button class=\\"funcbox-comment-submit\\" data-cancel=\\"1\\">Cancel</button>
                            <button class=\\"funcbox-comment-submit publish\\" data-submit=\\"1\\">Publish</button>
                        </div>
                    </form><ul class=\\"funcbox-comment-list\\">
                <!--;-->
            </ul><!--;--></div>"
`;

exports[`CommentComponent #render() 1`] = `
"<div id=\\"funcbox-comment\\"><h3 class=\\"funcbox-comment-heading\\">Comments</h3><div class=\\"funcbox-placeholder \\" data-showform=\\"1\\">
                        <span class=\\"funcbox-placeholder-text\\">
                            Write a comment...
                        </span>
                    </div><form class=\\"funcbox-comment-form hide\\">
                        <textarea class=\\"funcbox-comment-textarea\\" name=\\"body\\" placeholder=\\"Write a comment...\\"></textarea>
                        <div class=\\"funcbox-comment-footer\\">
                            <button class=\\"funcbox-comment-submit\\" data-cancel=\\"1\\">Cancel</button>
                            <button class=\\"funcbox-comment-submit publish\\" data-submit=\\"1\\">Publish</button>
                        </div>
                    </form><ul class=\\"funcbox-comment-list\\">
                <!--;-->
            </ul><!--;--></div>"
`;
